File: util-content/src/main/java/com/hendraanggrian/util/content/Themes.java
package com.hendraanggrian.util.content;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.support.annotation.AttrRes;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class Themes {

    private Themes() {
    }

    @Nullable
    public static CharSequence getText(@NonNull Context context, @AttrRes int attr) {
        return getText(context.getTheme(), attr);
    }

    @Nullable
    public static CharSequence getText(@NonNull Resources.Theme theme, @AttrRes int attr) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getText(0);
        } finally {
            a.recycle();
        }
    }

    @Nullable
    public static String getString(@NonNull Context context, @AttrRes int attr) {
        return getString(context.getTheme(), attr);
    }

    @Nullable
    public static String getString(@NonNull Resources.Theme theme, @AttrRes int attr) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getString(0);
        } finally {
            a.recycle();
        }
    }

    public static boolean getBoolean(@NonNull Context context, @AttrRes int attr, boolean defValue) {
        return getBoolean(context.getTheme(), attr, defValue);
    }

    public static boolean getBoolean(@NonNull Resources.Theme theme, @AttrRes int attr, boolean defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getBoolean(0, defValue);
        } finally {
            a.recycle();
        }
    }

    public static int getInt(@NonNull Context context, @AttrRes int attr, int defValue) {
        return getInt(context.getTheme(), attr, defValue);
    }

    public static int getInt(@NonNull Resources.Theme theme, @AttrRes int attr, int defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getInt(0, defValue);
        } finally {
            a.recycle();
        }
    }

    public static float getFloat(@NonNull Context context, @AttrRes int attr, float defValue) {
        return getFloat(context.getTheme(), attr, defValue);
    }

    public static float getFloat(@NonNull Resources.Theme theme, @AttrRes int attr, float defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getFloat(0, defValue);
        } finally {
            a.recycle();
        }
    }

    @ColorInt
    public static int getColor(@NonNull Context context, @AttrRes int attr, @ColorInt int defValue) {
        return getColor(context.getTheme(), attr, defValue);
    }

    @ColorInt
    public static int getColor(@NonNull Resources.Theme theme, @AttrRes int attr, @ColorInt int defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getColor(0, defValue);
        } finally {
            a.recycle();
        }
    }

    @Nullable
    public static ColorStateList getColorStateList(@NonNull Context context, @AttrRes int attr) {
        return getColorStateList(context.getTheme(), attr);
    }

    @Nullable
    public static ColorStateList getColorStateList(@NonNull Resources.Theme theme, @AttrRes int attr) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getColorStateList(0);
        } finally {
            a.recycle();
        }
    }

    public static float getDimension(@NonNull Context context, @AttrRes int attr, float defValue) {
        return getDimension(context.getTheme(), attr, defValue);
    }

    public static float getDimension(@NonNull Resources.Theme theme, @AttrRes int attr, float defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getDimension(0, defValue);
        } finally {
            a.recycle();
        }
    }

    public static int getDimensionPixelOffset(@NonNull Context context, @AttrRes int attr, int defValue) {
        return getDimensionPixelOffset(context.getTheme(), attr, defValue);
    }

    public static int getDimensionPixelOffset(@NonNull Resources.Theme theme, @AttrRes int attr, int defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getDimensionPixelOffset(0, defValue);
        } finally {
            a.recycle();
        }
    }

    public static int getDimensionPixelSize(@NonNull Context context, @AttrRes int attr, int defValue) {
        return getDimensionPixelSize(context.getTheme(), attr, defValue);
    }

    public static int getDimensionPixelSize(@NonNull Resources.Theme theme, @AttrRes int attr, int defValue) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getDimensionPixelSize(0, defValue);
        } finally {
            a.recycle();
        }
    }

    @Nullable
    public static Drawable getDrawable(@NonNull Context context, @AttrRes int attr) {
        return getDrawable(context.getTheme(), attr);
    }

    @Nullable
    public static Drawable getDrawable(@NonNull Resources.Theme theme, @AttrRes int attr) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getDrawable(0);
        } finally {
            a.recycle();
        }
    }

    @Nullable
    public static CharSequence[] getTextArray(@NonNull Context context, @AttrRes int attr) {
        return getTextArray(context.getTheme(), attr);
    }

    @Nullable
    public static CharSequence[] getTextArray(@NonNull Resources.Theme theme, @AttrRes int attr) {
        TypedArray a = theme.obtainStyledAttributes(new int[]{attr});
        try {
            return a.getTextArray(0);
        } finally {
            a.recycle();
        }
    }
}

File: src/com/hendraanggrian/kota/content/Themes.kt
package com.hendraanggrian.kota.content

import android.content.res.ColorStateList
import android.content.res.Resources
import android.content.res.TypedArray
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.support.annotation.AttrRes
import android.support.annotation.ColorInt
import android.support.annotation.Px

/**
 * Obtain single attribute and immediately recycle the [TypedArray].
 * Every extension functions have boolean as optional second param.
 * When set to true, the third param (default value, if any) can be ignored and will return non-null value.
 *
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */

@JvmOverloads
fun Resources.Theme.getText(@AttrRes id: Int, strict: Boolean = false): CharSequence? {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getText(0)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getString(@AttrRes id: Int, strict: Boolean = false): String? {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getString(0)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getBoolean(@AttrRes id: Int, strict: Boolean = false, def: Boolean = false): Boolean {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getBoolean(0, def)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getInt(@AttrRes id: Int, strict: Boolean = false, def: Int = 0): Int {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getInt(0, def)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getFloat(@AttrRes id: Int, strict: Boolean = false, def: Float = 0f): Float {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getFloat(0, def)
    } finally {
        a.recycle()
    }
}

@ColorInt
@JvmOverloads
fun Resources.Theme.getColor(@AttrRes id: Int, strict: Boolean = false, @ColorInt def: Int = Color.TRANSPARENT): Int {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getColor(0, def)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getColorStateList(@AttrRes id: Int, strict: Boolean = false): ColorStateList? {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getColorStateList(0)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getDimension(@AttrRes id: Int, strict: Boolean = false, defValue: Float = 0f): Float {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getDimension(0, defValue)
    } finally {
        a.recycle()
    }
}

@Px
@JvmOverloads
fun Resources.Theme.getDimensionPixelOffset(@AttrRes id: Int, strict: Boolean = false, def: Int): Int {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getDimensionPixelOffset(0, def)
    } finally {
        a.recycle()
    }
}

@Px
@JvmOverloads
fun Resources.Theme.getDimensionPixelSize(@AttrRes id: Int, strict: Boolean = false, def: Int): Int {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getDimensionPixelSize(0, def)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getDrawable(@AttrRes id: Int, strict: Boolean = false): Drawable? {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getDrawable(0)
    } finally {
        a.recycle()
    }
}

@JvmOverloads
fun Resources.Theme.getTextArray(@AttrRes id: Int, strict: Boolean = false): Array<CharSequence>? {
    val a = obtainStyledAttributes(intArrayOf(id))
    if (strict && !a.hasValue(0)) {
        throw NullPointerException()
    }
    try {
        return a.getTextArray(0)
    } finally {
        a.recycle()
    }
}